docs/window-shims-and-polyfills.md

```diff
@@ -0,0 +1,120 @@
+  /** Reference to SomeLibrary. */
+  SomeLibrary: any;
+  // ...
+};
+```
+
+Lets take `SomeLibrary` as an example. We just assign any to it, so that any reference in the code would not cause an TS compilation error. As we have properly used feature detection, in our application, (as shown below) the code that is actually using the library is _not_ called at all. However, Typescript is still compiling the code, and will error out if it's not available on the global scope.
+
+```typescript
+import { Component, ElementRef, Inject, OnInit, PLATFORM_ID } from '@angular/core';
+
+@Component({
+  selector: 'app-home',
+  templateUrl: './home.component.html',
+  styleUrls: ['./home.component.css'],
+})
+export class HomeComponent implements OnInit {
+  constructor(
+    private elmRef: ElementRef,
+    @Inject(PLATFORM_ID) private platformId: Object
+    ) {}
+
+  ngOnInit(): void {
+    /**
+     * We need the element anyway, and the nativeElement is
+     * only available in the browser, not during SSR!
+     * So we can can use that as feature detection.
+     */
+    const elm = this.elmRef.nativeElement as HTMLElement;
+    if (elm) {
+      /** somehow we made SomeLibrary available in the browser */
+      SomeLibrary.addExitingFunctionality(elm);
+    }
+    /** as an alternative you can check for the existence of the lib */
+    if (SomeLibrary) {
+      SomeLibrary.addExitingFunctionality(elm);
+    }
+    /**
+     * Or use isPlatformBrowser. Don't forget to inject the platformId then.
+     */
+    if (isPlatformBrowser(this.platformId)) {
+      SomeLibrary.addExitingFunctionality(elm);
+    }    
+  }
+}
+```
+
+When you have TypeScript typing error during compilation for SSR, start off with adding the missing type to the global decoration. the type `any` will do, but providing proper typing will help you discover problems sooner.
+
+### 2. Node shims 
+
+When your application is using some browser (or 3rth party) functionality that is not available in node, you might want to provide a stub function that as the same type signature, but isn't actually doing anything. I will take the Event function. It creates a new event. Still, there is no way you can trigger that during SSR. So its needs to pass compilation, but nothing more.
+
+```typescript
+global.Event = window.Event = (function () {} as unknown) as ((n) =>Event);
+```
+
+This will allow your app to compile and run during SSR. 
+
+> ## note
+> Perhaps you noticed the: `global.Event = window.Event = ...` in the above sample.  
+> We did that because in the browser, all the things on `window` are also available in the global namespace. As we don't have control over how our 3rth parties are using those, we need to provide both.
+
+### 3. node polyfills
+
+Node polyfills are used where there is no other way around getting your application up and running. Lets say we have a legacy library that uses a requestAnimationFrame in its init routine, and it will crash SSR if its not there.
```

Comment:
Node polyfills are used when there is no other way to get your application through the Universal compilation process. Let's look at an example of shimming for a legacy library that uses requestAnimationFrame in its init routine. This will crash during SSR if requestAnimationFrame is not present.